packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -1126,8 +1126,43 @@ abstract class State<T extends StatefulWidget> with Diagnosticable {
   /// ```
   ///
   /// It is an error to call this method after the framework calls [dispose].
-  /// You can determine whether it is legal to call this method by checking
+  /// You can determine whether it is legal to call [setState] by checking
   /// whether the [mounted] property is true.
+  ///
+  /// # Performance considerations
+  ///
+  /// Calls to [setState] cause a rebuild of the current widget and,
+  /// potentially, all of its children. Generally, [setState] calls should be
+  /// made on widgets with small subtrees in order to rebuild the least amount
```

Comment:
This is a little misleading. If you need to replace the entire content of your app, you will have to call setState on a widget that has a large subtree underneath it. And there is nothing wrong with it. If you call setState on the root of your app to update a counter widget somewhere at its leaf, that's probably a problem. In other words: you should be pushing state to its leaves where possible - which naturally will lead to an architecture where setState only rebuilds what's necessary.

Take a look at the "Performance considerations" at https://master-api.flutter.dev/flutter/widgets/StatefulWidget-class.html it explains this in a pretty neutral way.